src/plots/linePlot.ts

```diff
@@ -61,6 +140,136 @@ export module Plots {
       return new Plottable.Drawers.Line(dataset);
     }
 
+    protected _extentsForProperty(property: string) {
+      let extents = super._extentsForProperty(property);
+
+      if (!this._autorangeSmooth) {
+        return extents;
+      }
+
+      if (this.autorangeMode() !== property) {
+        return extents;
+      }
+
+      if (this.autorangeMode() !== "x" && this.autorangeMode() !== "y") {
+        return extents;
+      }
+
+      let edgeIntersectionPoints = this._getEdgeIntersectionPoints();
+      let includedValues: number[];
+      if (this.autorangeMode() === "y") {
+        includedValues = edgeIntersectionPoints.left.concat(edgeIntersectionPoints.right).map((point) => point.y);
+      } else { // === "x"
+        includedValues = edgeIntersectionPoints.top.concat(edgeIntersectionPoints.bottom).map((point) => point.x);
+      }
+
+      let maxIncludedValue = Math.max.apply(this, includedValues);
+      let minIncludedValue = Math.min.apply(this, includedValues);
+
+      extents = extents.map((extent) => {
+        // When no end-points are in the viewport
+        if (extent.length === 0) {
+          return [minIncludedValue, maxIncludedValue];
+        }
+        return [Math.min(minIncludedValue, extent[0]), Math.max(maxIncludedValue, extent[1])];
+      });
+
+      return extents;
+    }
+
+    private _getEdgeIntersectionPoints(): EdgeIntersection {
+      if (!(this.y().scale instanceof QuantitativeScale && this.x().scale instanceof QuantitativeScale)) {
+        return {
+          top: [],
+          bottom: [],
+          left: [],
+          right: []
+        };
+      }
+
+      let yScale = <QuantitativeScale<number>>this.y().scale;
+      let xScale = <QuantitativeScale<any>>this.x().scale;
+
+      let intersectionPoints: EdgeIntersection = {
+        top: [],
+        bottom: [],
+        left: [],
+        right: []
+      };
+      let leftX = xScale.scale(xScale.domain()[0]);
+      let rightX = xScale.scale(xScale.domain()[1]);
+      let bottomY = yScale.scale(yScale.domain()[0]);
+      let topY = yScale.scale(yScale.domain()[1]);
+
+      this.datasets().forEach((dataset) => {
```

Comment:
I think you meant the other way round. Reduce from `O(n)` to `O(log(n))` by assuming the points are ordered and doing 2 binary searches.

1. From previous feedback from QE and PM, it seemed like such assumptions might not be reasonable. It is not clear from the method name that the points have to be ordered, and it is very hard to explode gracefully if the points are unordered, so the user might not know what is going on.

2. In-memory operations are not as costly as DOM manipulation or rendering. (I mean by 1-2 orders of magnitude). I'd prefer not to optimize early, but I will profile this and see if you're right.